From 4bf70604c2303caba4a6d975bf504c4d5e23095b Mon Sep 17 00:00:00 2001
From: Andrew de los Reyes <adlr@chromium.org>
Date: Tue, 2 Jul 2019 15:54:16 -0700
Subject: [PATCH] FPDFText_GetIndexForOffset

Adds and API to get the text caret offset for a given floating point
position into a text object.
---
 core/fpdfapi/page/cpdf_textobject.cpp | 18 ++++++++++++++++++
 core/fpdfapi/page/cpdf_textobject.h   |  1 +
 fpdfsdk/fpdf_edittext.cpp             |  8 ++++++++
 public/fpdf_edit.h                    |  9 +++++++++
 4 files changed, 36 insertions(+)

diff --git a/core/fpdfapi/page/cpdf_textobject.cpp b/core/fpdfapi/page/cpdf_textobject.cpp
index cc0827812..4071c6594 100644
--- a/core/fpdfapi/page/cpdf_textobject.cpp
+++ b/core/fpdfapi/page/cpdf_textobject.cpp
@@ -194,6 +194,24 @@ float CPDF_TextObject::GetCharWidth(uint32_t charcode) const {
   return pCIDFont->GetVertWidth(CID) * fontsize;
 }
 
+int CPDF_TextObject::GetIndexForOffset(float xpos) const {
+  xpos -= m_Rect.left;
+  int ret = 0;
+  for (size_t i = 0; i < m_CharCodes.size(); i++) {
+    uint32_t charcode = m_CharCodes[i];
+    if (charcode == CPDF_Font::kInvalidCharCode)
+      continue;
+    float width = GetCharWidth(charcode);
+    float charstart = 0;
+    if (i > 0)
+      charstart = m_CharPos[i - 1];
+    if (xpos < (charstart + width / 2))
+      break;
+    ret++;
+  }
+  return ret;
+}
+
 CPDF_Font* CPDF_TextObject::GetFont() const {
   return m_TextState.GetFont();
 }
diff --git a/core/fpdfapi/page/cpdf_textobject.h b/core/fpdfapi/page/cpdf_textobject.h
index 9ed5ff7ff..56ffa446d 100644
--- a/core/fpdfapi/page/cpdf_textobject.h
+++ b/core/fpdfapi/page/cpdf_textobject.h
@@ -45,6 +45,7 @@ class CPDF_TextObject final : public CPDF_PageObject {
   void GetCharInfo(size_t index, uint32_t* charcode, float* kerning) const;
   void GetCharInfo(size_t index, CPDF_TextObjectItem* pInfo) const;
   float GetCharWidth(uint32_t charcode) const;
+  int GetIndexForOffset(float xpos) const;
 
   CFX_PointF GetPos() const { return m_Pos; }
   CFX_Matrix GetTextMatrix() const;
diff --git a/fpdfsdk/fpdf_edittext.cpp b/fpdfsdk/fpdf_edittext.cpp
index c7448141e..7ac862312 100644
--- a/fpdfsdk/fpdf_edittext.cpp
+++ b/fpdfsdk/fpdf_edittext.cpp
@@ -465,6 +465,14 @@ FPDFText_SetText(FPDF_PAGEOBJECT text_object, FPDF_WIDESTRING text) {
   return true;
 }
 
+FPDF_EXPORT int FPDF_CALLCONV
+FPDFText_GetIndexForOffset(FPDF_PAGEOBJECT text_object, double xpos) {
+  CPDF_TextObject* pTextObj = CPDFTextObjectFromFPDFPageObject(text_object);
+  if (!pTextObj)
+    return -1;
+  return pTextObj->GetIndexForOffset(static_cast<float>(xpos));
+}
+
 FPDF_EXPORT FPDF_FONT FPDF_CALLCONV FPDFText_LoadFont(FPDF_DOCUMENT document,
                                                       const uint8_t* data,
                                                       uint32_t size,
diff --git a/public/fpdf_edit.h b/public/fpdf_edit.h
index bd8461e2a..38f1f835b 100644
--- a/public/fpdf_edit.h
+++ b/public/fpdf_edit.h
@@ -1214,6 +1214,15 @@ FPDFPageObj_NewTextObj(FPDF_DOCUMENT document,
 FPDF_EXPORT FPDF_BOOL FPDF_CALLCONV
 FPDFText_SetText(FPDF_PAGEOBJECT text_object, FPDF_WIDESTRING text);
 
+// Get the nearest index between two characters for a given x position.
+//
+// text_object - handle to the text object.
+// xpos        - x position in page coordinates.
+//
+// Returns a number in range [0, length] or -1 on error.
+FPDF_EXPORT int FPDF_CALLCONV
+FPDFText_GetIndexForOffset(FPDF_PAGEOBJECT text_object, double xpos);
+
 // Returns a font object loaded from a stream of data. The font is loaded
 // into the document.
 //
-- 
2.11.0

